### ISSUE TITLE
Historical rack aware (or group aware) data replication 

### ISSUE THREAD

**Issue by capistrant** [apache/druid#9816]
***Note***: I use the term "Rack" throughout this proposal because I am speaking from my experience deploying Druid on bare metal servers on prem with my employer. It may be more wise if we think of this as a logical grouping of historical nodes that in my case happens to be physical racks. That way in a virtualized deployment where there may not be any concept of a physical rack, admins could still balance segments based on the logical groupings of historical services in order to capture the same value that I lay out below.

### Motivation

Right now we are only guaranteed that replicas are not located on the same node. However, as you scale a cluster out to many nodes it becomes time intensive to rolling restart Historical services. For instance, if you have 2x replication in a single tier, you have to restart only one node at a time in order to maintain data availability. What we'd like to do is have Historical nodes be assigned an attribute to specify what rack they are on. Then if, rack awareness is turned on, Druid makes a best effort to have segments located across multiple racks. This would allow batch restarts for Historical services in large clusters with little to no unavailable data (this would be best effort cross rack replication, so there is still a chance you have all replicas on one rack).

### Proposed changes

I have a POC going in my own fork that I have been working on. I will link a diff with master [here](https://github.com/apache/druid/compare/master...capistrant:druid-rack-awareness) in addition to the notes below.

**Configs**
Two new configs, one to specify the rack of a server. Another boolean flag to specify whether or not we are enabling rack aware replication. Druid server objects would then be instantiated with these attributes.

* `druid.rack`
* `druid.enableRackAwareness`

**Cost Balancer**

In my POC, I left the random balancer alone. I only implemented the optional rack based replication for the cost balancer. I decided it would be simplest to keep the rack based decision making outside of the cost calculations. Therefore, the cost calculations are made for a segment as normal and then when choosing the best server to put a segment on or drop a segment from, there are considerations for rack distribution. The goal being that we want to bias towards increasing or maintaining rack distribution for a segment. Therefore a server that may have the lowest cost of all candidates when deciding where to place a segment might not be chosen. Instead a higher cost server could be chosen if it means increasing rack distribution of a segment.

### Rationale

I believe that my initial approach of keeping the rack distribution logic outside of the cost calculations is best because it is simplest. In order to fold rack decisions into the cost calculations, I imagine you'd have to maintain some large thread safe data structure that contains information on what rack a segment currently exists on. That way cost calculations could factor in whether or not the candidate server is on a rack currently serving the segment. I could be wrong in stating that the complexity and cost of this seems to rule it out as the right solution path, and I am certainly willing to explore it if deemed worthwhile.

### Operational impact

This section should describe how the proposed changes will impact the operation of existing clusters. It should answer questions such as:

- Is anything going to be deprecated or removed by this change? How will we phase out old behavior?
- Is there a migration path that cluster operators need to be aware of?
- Will there be any effect on the ability to do a rolling upgrade, or to do a rolling _downgrade_ if an operator wants to switch back to a previous version?

The config defaults would result in the cost and caching cost balancers having the same behavior as they already do. The implementation would also ensure the performance of the balancer would be the same if this feature is not turned on. If someone decides to turn the feature on, their cluster would start rebalancing and it could lead to interesting balance across the cluster if there is not some initial analysis done on the cluster architecture and whether or not there is a good enough distribution of servers across racks to result in acceptable data balance across the cluster. If you were to downgrade, your balancer would go back to not using rack information to assign segments so you'd go through a phase where your cluster could get into some interesting balance states depending on the distribution of servers across racks.

### Test plan (optional)

The balancer class seems like something that could easily have a simulation written for it where we have multiple cluster architectures and run simulations on them with N segments. You could balance the cluster with rack awareness off and then turn it on and track the re-distribution of segments using rack information to make decisions. 

### Future work (optional)

N/A

**Comment by himanshug** [comment]
Isn't this doable using multiple historical tiers ... e.g.
some historicals with `druid.tier=A`
some historicals with `druid.tier=B`

and then using `tieredReplicants` in the loadRule e.g.

```
  "tieredReplicants": {
      "A": 1,
      "B" : 1
  }
```

**Comment by capistrant** [comment]
for a case where number of groups or racks = number of replicas desired, sure. But a large enterprise cluster will have 10s of racks and will not want a replica on every rack due to cost concerns. And trying to automatically generate load rules that properly balance datasources across those groups by selecting tiers would be burdensome and likely hard to do well.

**Comment by himanshug** [comment]
> for a case where number of groups or racks = number of replicas desired, sure. But a large enterprise cluster will have 10s of racks and will not want a replica on every rack due to cost concerns. 

that should be possible I think.

say, you have 4 racks/groups/tiers ... A, B, C, D

with config...

```
  "tieredReplicants": {
      "A": 1,
      "B" : 2
  }
```

number of replicants != number of tiers , and not every replica is on each tier.

or maybe I misunderstood above statement.

> And trying to automatically generate load rules that properly balance datasources across those groups by selecting tiers would be burdensome and likely hard to do well.

Yeah, things could possibly be simplified but use of tiers does let you achieve the end goal. Just wanted to point out the existing alternative and also it would be nice if tier based grouping could be enhanced to take care of additional use cases or simplification instead of introducing another way of grouping historical.

**Comment by ArvinZheng** [comment]
@himanshug ,
Yes, use multiple tiers can be an alternative for some cases, but it doesn't work in the case that you want to keep the same number of active replicants regardless of how many racks/zones that you have in the cluster. For example, there will be only one active replicant if rack B goes down in your example.

**Comment by himanshug** [comment]
that is true and I am suggesting if possible, that enhancement could be made to tier concept itself rather than introducing another thing to do the grouping 

just for illustration, having the config be something like e.g.

```
"numReplicas": 3
"tiers": [ "A", "B", "C" ]     //optional and defaults to all/any available tiers
```

**Comment by ArvinZheng** [comment]
I see your point, but people (like me) may have gotten used to leverage tiers to split data into cold and hot groups and allocation different resources for cold and hot tiers. While rack awareness is a different concept which we want to make a tier more tolerant for rack/zone failures. I vote to use different configs/terms for introducing this feature.

**Comment by dougbyrne** [comment]
In the case of cloud hosting, I generally think of "rack" corresponding to "availability-zone." My tiers and AZs will overlap in ways that don't make sense in the context of physical racks.

We are in the process of adding on-demand and spot tiers. I can specify the number of replicas for each tier to mitigate against the possibility of spot instances becoming unavailable. I also need to mitigate against AZ failure. Each AZ can contain a mix of on-demand and spot nodes. I would like the distribution of replicas to also be aware of AZ, even across the tiers.

I'm not sure that having a tier for each combination of AZ and pricing class will satisfy both conditions.

_AWS Jargon: On-demand instances are the "normal" type. Spot instances are lower cost, but can be pre-empted by requests for on-demand instances._

**Comment by ArvinZheng** [comment]
@dougbyrne ,
"rack" doesn't have to be a physical rack here, in your case the value of `druid.rack` will be AZ.

The goal of this PR is to make sure Druid distributes replications between racks/AZs within a tier, so to achieve this, you should make sure every tier in your Druid cluster has historical nodes run in multiple AZs. 
@capistrant correct me if I am wrong.
> I would like the distribution of replicas to also be aware of AZ, even across the tiers.

**Comment by dougbyrne** [comment]
Yes, that's what I said, in the context of cloud hosting, a "rack" is a AZ. I'm trying to illustrate why tiers do not satisfy my distribution requirements alone.

Lets I have tier 1 in rack A, rack B and rack C, and tier 2 also in racks A, B, and C. Then I configure tier 1 to have one 1 replica, and tier 2 to have 2 replicas. If my tier 1 replica was in rack A, and my tier 2 replicas were in racks A and B, that would not be acceptable to me. I should not have multiple replicas in the same rack when possible.

I'm not sure extending tiers will address this, as tiers and rack are different dimensions and can overlap.

**Comment by ArvinZheng** [comment]
hmm, sorry @dougbyrne , I'm not sure if I understood your concern.

In the example that you posted, tier 1 wouldn't be beneficial by the rack awareness proposal since you have only 1 replica in that tier, so anyway once you lose one rack, a portion of data will be unavailable until they are redistributed to other rack. But the data in tier 2 are supposed to be able to tolerate at least one zone failure by this proposal since the segment distribution will prioritize rack first, and if you configure tier 2 to have 3 replicas, then we would expect tier 2 is able to tolerate 2 zones being unavailable.

**Comment by dougbyrne** [comment]
In my example, I'm loading the same data in both Tier 1 and Tier 2. I have a total of 3 replicas. The rack awareness needs to cross tier boundaries.

**Comment by ArvinZheng** [comment]
Ah, I see, may I ask one more question of your use case? What's the purpose to set up 2 tiers to serve the same data?

**Comment by dougbyrne** [comment]
I have one tier running on On-Demand instances, another tier running on spot (preemptable) instances. I want to ensure that there is at least one replica for our hottest data that will not be preempted.

**Comment by ArvinZheng** [comment]
I see, would it be possible/acceptable to serve the data in the same tier, and use different AZs for on-demand and spot instances? So that you don't have to use multiple tiers to serve the same data and let Druid to replicate the segments to different AZs and make sure the cluster is able to tolerate AZ failures (assuming we have this rack awareness feature implemented and enabled)?

**Comment by dougbyrne** [comment]
That would not be ideal. The availability of spot instances is different per AZ, so we want to have as many AZs available to our spot tier. We also need to have multiple AZs available in the on-demand tier to mitigate the possibility of a AZ failure.

I think you raise a good point that tiers are not the best way to solve this problem. Maybe instead of having just rack diversity, it should be more generic. Rack and Zone and Price Class are all dimensions you might want to diversify across.

Also, our load rules are a bit more complicated. The most recent data are in both the on-demand tier and the hot-spot tier. The nxt most recent data is only in that same hot-spot tier. The remaining data is loaded on a cold-spot tier. We also only want to have one replica in the on-demand tier, while we might have more in the spot tier.

**Comment by capistrant** [comment]
@dougbyrne  @ArvinZheng @himanshug 

Hi all. Sorry for not being active in the discussion last week. I have revisited my initial proposal and came up with a new POC branch that I think is better than the first attempt. [code link here](https://github.com/apache/druid/compare/master...capistrant:guild-replication-poc)

I walked away from the use of the term "rack". For now I have used the term "guild".... but naming is not the issue at hand here, we can easily change that. At the end of the day, a guild is just a group. If an operator so chooses they can assign every historical to a guild and then tell the coordinator that guild replication is on. This means that the coordinator should make a best effort to load replicants onto guilds that are not already used whenever possible. However, if a replicant has to go onto a used guild, the coordinator will do this because having N replicants on one guild is better than N-1 replicants in the cluster when there are no unused guilds to place a segment. A guild can be anything. For my team, it will be a physical rack. For Doug, it may be an AZ or an AZ:price class combination. 

I am not a very big fan of layering into this into tiering. Tiering as is seems so much more like a query routing/prioritization/isolation feature. I want my router/broker to make query choices based on my tiers. I do not want them to care about my guilds. A segment being on guild A vs guild Z means nothing in terms of which should be queried. It is just a way to help operators improve operational simplicity and uptime reliability. If we were going to roll this into tiering, I fear that we would create one very confusing beast rather than two separate and generally simple ideas.

My aforementioned link to my POC code still needs heavy testing and iteration to work out the kinks (I'm sure there are many). But I think it is a good anchor point to base future discussions on. I can take the time to update my base proposal with new details to this implementation if people think that is a good idea. Please let me know! In the meantime I will continue working on testing/validation. I really want to look into the balancer profiler and see what that is all about and if I can turn it into some type of simulation of a cluster at scale turning on guild awareness.

**Comment by himanshug** [comment]
I haven't seen the code but from a quick read of comments above, I see that proposed new configuration would cover multiple segment distribution related use cases , so +1 on that.

**Comment by capistrant** [comment]
I've updated the proposal to reflect the latest implementation I have used for my POC. Will be adding some more changes for testing this week I hope.

**Comment by ngaugler** [comment]
Has any progress been made on anything over the years or is it still impossible to appropriately redundantly balance data?

**Comment by abhishekagarwal87** [comment]
Is the tiering not sufficient for the use case you have? It won't solve for all scenarios but is good enough for many of them.

**Comment by capistrant** [comment]
> Has any progress been made on anything over the years or is it still impossible to appropriately redundantly balance data?

This has remained stalled, but I plan on picking up the effort again sometime this fall if it remains open
